File: recominer-extractor/src/main/resources/scripts/classification.py
porco1 = [1, 1, 0]
porco2 = [1, 1, 0]
porco3 = [1, 1, 0]
cachorro4 = [1, 1, 1]
cachorro5 = [0, 1, 1]
cachorro6 = [0, 1, 1]

dados = [porco1, porco2, porco3, cachorro4, cachorro5, cachorro6]

marcacoes = [1, 1, 1, -1, -1, -1]

from numpy import genfromtxt
train = genfromtxt('/home/kuroda/NetBeansProjects/extractor/generated/avro/259/1281/256/train.csv', skip_header=1, names="file1Id, file1, file2Id, file2, id, issueId, commitId, issueKey, issueType, issuePriority, issueAssignedTo, issueSubmittedBy, commenters, devCommenters, issueAge, wordinessBody, wordinessComments, comments, networkId, networkIssueId, networkCommitId, btwMdn, clsMdn, dgrMdn, efficiencyMdn, efvSizeMdn, constraintMdn, hierarchyMdn, size, ties, density, diameter, commitMetricId, commitMetricCommitId, revision, committer, fileMetricId, fileId, committers, commits, fileAge, addedLines, deletedLines, changedLines, cochanged", delimiter=';', dtype=[('file1Id', '<i4'), ('file1', '|S1024'), ('file2Id', '<i4'), ('file2', '|S1024'), ('id', '<i4'), ('issueId', '<i4'), ('commitId', '<i4'), ('issueKey', '|S32'), ('issueType', '|S32'), ('issuePriority', '|S32'), ('issueAssignedTo', '|S128'), ('issueSubmittedBy', '|S128'), ('commenters', '<i4'), ('devCommenters', '<i4'), ('issueAge', '<i4'), ('wordinessBody', '<i4'), ('wordinessComments', '<i4'), ('comments', '<i4'), ('networkId', '<i4'), ('networkIssueId', '<i4'), ('networkCommitId', '<i4'), ('btwMdn', '<f8'), ('clsMdn', '<f8'), ('dgrMdn', '<f8'), ('efficiencyMdn', '<f8'), ('efvSizeMdn', '<f8'), ('constraintMdn', '<f8'), ('hierarchyMdn', '<f8'), ('size', '<i4'), ('ties', '<i4'), ('density', '<f8'), ('diameter', '<i4'), ('commitMetricId', '<i4'), ('commitMetricCommitId', '<i4'), ('revision', '|S128'), ('committer', '|S128'), ('fileMetricId', '<i4'), ('fileId', '<i4'), ('committers', '<i4'), ('commits', '<i4'), ('fileAge', '<i4'), ('addedLines', '<i4'), ('deletedLines', '<i4'), ('changedLines', '<i4'), ('cochanged', '<i4')])
print(train)

#from sklearn.naive_bayes import MultinomialNB
from sklearn.ensemble import RandomForestClassifier

#modelo = MultinomialNB()
modelo = RandomForestClassifier(n_estimators=10)

modelo.fit(dados, marcacoes)

misterioso1 = [1, 1, 1]
misterioso2 = [1, 0, 0]
misterioso3 = [0, 0, 1]

teste = [misterioso1, misterioso2, misterioso3]

marcacoes_teste = [-1, 1, -1]

resultado = modelo.predict(teste)

print(resultado)

File: recominer-extractor/src/main/resources/scripts/classification.R
options(echo=TRUE) # if you want see commands in output file / , warn=2

library("caret")
library("mlbench")
library("AppliedPredictiveModeling")
library("randomForest")
library("e1071")

set.seed(10)

# Reading parameters passed
args <- commandArgs(trailingOnly = TRUE)
print(args)

workingDirectory <- args[1]
projectName <- args[2]
commit <- args[3]
file <- args[4]
cochange <- args[5]

# Folder structure: PROJECT/ISSUE/COMMIT/FILE
resultFolder <- paste(workingDirectory, projectName, commit, file, sep = "/")
print(paste("Result folder: ", resultFolder))
setwd(resultFolder)

sub.folders <- list.dirs(resultFolder, recursive=FALSE)

for(cochangeSubfolder in sub.folders) {
  trainFile <- paste(cochangeSubfolder, "train.csv", sep = "/")
  testFile <- paste(resultFolder, "test.csv", sep = "/")
  
  print(paste("Reading data from", resultFolder))
  train <- read.csv(trainFile, sep=";", na.strings=c("NA",""))
  test <- read.csv(testFile, sep=";", na.strings=c("NA",""))
  
  file1Id <- train[1,1]
  file1 <- train[1,2]
  file2Id <- train[1,3]
  file2 <- train[1,4]
  
  #cochanged<-train[,c("cochanged")]
  train<-train[,c("commenters", "devCommenters", "issueAge", "wordinessBody", "wordinessComments", "comments", "btwMdn", "clsMdn", "dgrMdn", "efficiencyMdn", "efvSizeMdn", "constraintMdn", "hierarchyMdn", "size", "ties", "density", "diameter", "committers", "commits", "fileAge", "addedLines", "deletedLines", "changedLines", "cochanged")]
  test<-test[,c("commenters", "devCommenters", "issueAge", "wordinessBody", "wordinessComments", "comments", "btwMdn", "clsMdn", "dgrMdn", "efficiencyMdn", "efvSizeMdn", "constraintMdn", "hierarchyMdn", "size", "ties", "density", "diameter", "committers", "commits", "fileAge", "addedLines", "deletedLines", "changedLines", "cochanged")]
  
  
  # Transforming the class (cochanged == 1 to "Changed", otherwise "Not changed")
  train$cochanged <- as.factor(ifelse(train$cochanged==1, "C", "N"))
  
  # Check if has two factors
  if (length(levels(train$cochanged)) == 2) {
  
    print(paste("Training for file", file1Id, "and", file2Id))
    rf_model <- train(cochanged ~ ., method="rf",  tuneLength = 2, trControl=trainControl(method = "boot"), data=train, importance = FALSE)
    
    # Removing instances with information that no exists in test dataset
    # For example, if in test dataset has an instance that issueType is Task
    # and in train does not exist a issueType Task, then ommit the instance.
    test<-na.omit(test)
  
    print("Predicting...")
  
    # Has obs. in test?
    if (nrow(test) > 0) {
      predicted <- predict(rf_model, test)
      # gets the first one (and only one) probability for predicted class (C or N)
      probability <- predict(rf_model, test, type = "prob")[1, predicted]
    } else {
      predicted <- NA
      probability <- NA
    }
  } else {
    print(paste("Train file", trainFile, "does not have two factors for training."))
    predicted <- NA
    probability <- NA
  }
  output <- data.frame(file2Id, file2, predicted, probability, stringsAsFactors=TRUE)
  write.table(output, "resultsTest.csv", append=TRUE, eol = "\n", sep=";", col.names = F, row.names = F)  
}